File: src/python/problems/SpiralOrder2.py
# https://leetcode.com/problems/spiral-matrix-ii/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class SpiralOrder2:
    def Solution(self, n: int) -> list[list[int]]:
        martix = [[0] * n for _ in range(n)]

        x, y = 0, 0
        dx, dy = 1, 1
        count = 1
        target = n*n

        while count <= target:
            martix[y][x] = count
            count += 1

            if y == dy - 1 and x < n - dx:
                x += 1
            elif x == n - dx and y < n - dy:
                y += 1
            elif x > dx - 1:
                x -= 1
            elif y > dy:
                y -= 1
                if y == dy and x == dx - 1:
                    dx += 1
                    dy += 1

        return martix



SpiralOrder2().Solution(3)

File: src/csharp/Problems/SpiralOrder2.cs
//https://leetcode.com/problems/spiral-matrix-ii/

namespace LeetCode.Problems;

public sealed class SpiralOrder2 : ProblemBase
{
    [Theory]
    [ClassData(typeof(SpiralOrder2))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.Param(3).Result2dArray("[[1,2,3],[8,9,4],[7,6,5]]"))
          .Add(it => it.Param(4).Result2dArray("[[1,2,3,4],[12,13,14,5],[11,16,15,6],[10,9,8,7]]"))
          .Add(it => it.Param(5).Result2dArray("[[1,2,3,4,5],[16,17,18,19,6],[15,24,25,20,7],[14,23,22,21,8],[13,12,11,10,9]]"))
          .Add(it => it.Param(6).Result2dArray("[[1,2,3,4,5,6],[20,21,22,23,24,7],[19,32,33,34,25,8],[18,31,36,35,26,9],[17,30,29,28,27,10],[16,15,14,13,12,11]]"))
          .Add(it => it.Param(1).Result2dArray("[[1]]"));

    private int[][] Solution(int n)
    {
        var matrix = Enumerable.Repeat(n, n).Select(it => new int[it]).ToArray();

        var x = 0;
        var y = 0;
        var dy = 1;
        var dx = 1;
        var count = 1;
        var target = n * n;
        while(count <= target)
        {
            matrix[y][x] = count++;

            if (y == dy - 1 && x < n - dx)
            {
                x++;
            }
            else if (x == n - dx && y < n - dy)
            {
                y++;
            }
            else if (x > dx - 1)
            {
                x--;
            }
            else if (y > dy)
            {
                y--;
                if (y == dy && x == dx - 1)
                {
                    dx++;
                    dy++;
                }
            }
        }

        return matrix;
    }
}